leaflet/index.d.ts

```diff
@@ -817,67 +817,72 @@ declare namespace L {
          * and/or coordsToLatLng functions if provided as options.
          */
-        geometryToLayer(featureData: GeoJSON.Feature<GeoJSON.GeometryObject>, options?: GeoJSONOptions): Layer;
+        static geometryToLayer(featureData: GeoJSON.Feature<GeoJSON.GeometryObject>, options?: GeoJSONOptions): Layer;
 
         /**
          * Creates a LatLng object from an array of 2 numbers (longitude, latitude) or
          * 3 numbers (longitude, latitude, altitude) used in GeoJSON for points.
          */
-        coordsToLatLng(coords: [number, number]): LatLng;
-
-        coordsToLatLng(coords: [number, number, number]): LatLng;
+        static coordsToLatLng(coords: [number, number] | [number, number, number]): LatLng;
 
         /**
          * Creates a multidimensional array of LatLngs from a GeoJSON coordinates array.
          * levelsDeep specifies the nesting level (0 is for an array of points, 1 for an array of
          * arrays of points, etc., 0 by default).
          * Can use a custom coordsToLatLng function.
          */
-        coordsToLatLngs(coords: Array<number>, levelsDeep?: number, coordsToLatLng?: (coords: [number, number] | [number, number, number]) => LatLng): LatLng[]; // Not entirely sure how to define arbitrarily nested arrays
+        static coordsToLatLngs(
+            coords: any[],
+            levelsDeep?: number,
+            coordsToLatLng?: (coords: [number, number] | [number, number, number]) => LatLng): any[]; // Using any[] to avoid artificially limiting valid calls
 
         /**
          * Reverse of coordsToLatLng
          */
-        latLngToCoords(latlng: LatLng): [number, number] | [number, number, number];
+        static latLngToCoords(latlng: LatLng): number[];
 
         /**
          * Reverse of coordsToLatLngs closed determines whether the first point should be
          * appended to the end of the array to close the feature, only used when levelsDeep is 0.
          * False by default.
          */
-        latLngsToCoords(latlngs: Array<LatLng>, levelsDeep?: number, closed?: boolean): [number, number] | [number, number, number];
+        static latLngsToCoords(latlngs: any[], levelsDeep?: number, closed?: boolean): any[];  // Using any[] to avoid artificially limiting valid calls
 
         /**
          * Normalize GeoJSON geometries/features into GeoJSON features.
          */
-        asFeature(geojson: GeoJSON.GeometryObject): GeoJSON.Feature<GeoJSON.GeometryObject>;
+        static asFeature(geojson: GeoJSON.GeometryObject): GeoJSON.Feature<GeoJSON.GeometryObject>;
+
+        static asFeature(geojson: GeoJSON.Feature<GeoJSON.GeometryObject>): GeoJSON.Feature<GeoJSON.GeometryObject>;
+
+    }
+
+    /**
+     * Represents a GeoJSON object or an array of GeoJSON objects.
+     * Allows you to parse GeoJSON data and display it on the map. Extends FeatureGroup.
+     */
+    export interface GeoJSON extends FeatureGroup {
```

Comment:
BLUF: We can't easily get rid of the GeoJSON interface. And, leaving them separate actually provides exactly the functionality we want in this context.


**Long Version:**

Why it would be hard:
To get rid of the GeoJSON interface, we'd need to move the remaining methods over to the class. But, then we'd also need to make sure the GeoJSON class was implementing the FeatureGroup interface. But since there are no other declared classes, we can't just have GeoJSON extend FeatureGroup, we'd have to actually add all the expected methods to the GeoJSON class.

Leaflet uses factories for everything, so the typings generally don't need to define any constructors or classes. In this case, we needed to add the class so we could add static methods (since you can't declare static methods on an interface). Since we don't add a public constructor on the class, you still can't create a GeoJSON instance apart from the factory methods. But now you can call the static methods on the class.

Why this approach is ok (I think):
In TS it seems that interface and class definitions of the same name are merged by the compiler. This means that the 'GeoJSON' type that is returned by the factory method that creates a GeoJSON object is returning the union of the GeoJSON class and interface methods. In our case specifically the only impact that the merge has is to make it so that we can reference the type "GeoJSON" with all of the interface methods, while also directly invoking static methods on the "GeoJSON" class (eg. GeoJSON.doStaticStuff()) - which is exactly the behavior we want.

Let me know if you have an alternative approach you'd prefer.